File: server/src/middlewares/duration.ts
import { NotFoundError } from '@/errors';
import { validator } from 'hono/validator';

const query = validator('query', (value, _) => {
    const nodeId = value['nodeId'];
    if (!nodeId) throw new NotFoundError('Missing required query `nodeId`!');
    return {
        nodeId: parseInt(nodeId as string),
    };
});

const payload = validator('json', (value: TDurationPayload, _) => {
    if (!value.timeOnTask) throw new NotFoundError('missing `timeOnTask`');
    if (!value.startFrom) throw new NotFoundError('missing `startFrom`');
    if (!value.endAt) throw new NotFoundError('missing `endAt`');
    if (!value.taskId) throw new NotFoundError('missing `taskId`');

    const data: TDuration = {
        ...value,
        startFrom: new Date(`${value.startFrom}`),
        endAt: new Date(`${value.endAt}`),
    };
    return {
        data,
    };
});

export const durationValidator = {
    query,
    payload,
};

File: server/internal/models/duration.go
package models

import "time"

type Duration struct {
	Id          int       `json:"id,omitempty" db:"id"`
	TimeOnTask  int       `json:"timeOnTask" validate:"required" db:"time_on_task"`
	StartFrom   time.Time `json:"startFrom" validate:"required" db:"start_from"`
	EndAt       time.Time `json:"endAt" validate:"required" db:"end_at"`
	Description string    `json:"description" db:"description"`
	TaskId      int       `json:"taskId" validate:"required" db:"task_id"`
}
